README.md
# pr2_delivery #

This is a quick package put together to demonstrate a simple delivery task with the PR2.

It is written as an "action" server.

One run through the action does:

 * Navigate to the get-object pose
 * Put the right arm out and open the gripper
 * Wait for something to be put in the gripper (wait for gripper-wiggle)
 * Close gripper
 * If gripper closes too far, open again and repeat waiting for object
 * Tuck arms to a tucked-with-object pose
 * Navigate to the give-object pose
 * Put the right arm out to hand off the object
 * Wait for gripper-wiggle
 * Open gripper
 * Tuck arms to regular tuck
 * Navigate to "home" pose

This sequence is defined in src/pr2_delivery/DeliverServer.py.

## Programs ##

 * **scripts/deliver**: Runs the DeliverServer action server.

 * **test/test.py**: sends an action goal asking it to make a delivery.
   * The action goal contains the 3 poses: get-object, give-object, and home.
   * *This is a file you will need to copy and modify for use in a new place.*

 * **scripts/print_right_arm_pose**: repeatedly prints the pose of the right arm in the same order that is used inside DeliveryServer.py.  Use this with (or without) the run-stop pressed to record new arm poses if you need to.

 * **test/watch_robot_base_pose.sh**: repeatedly prints the pose of the robot base.  This can be convenient for recording new poses for get-object, give-object, and home.

## Launch files ##

 * **launch/deliver_server.launch**: runs scripts/deliver and a bunch of support nodes which it depends on.
   * Does *not* run map_server.

 * **launch/deliver_server_wg.launch**: runs deliver_server.launch and map_server with a map of Willow Garage.
   * *This is a file you will need to copy and modify for use in a new place.*

## Instructions for use ##

### Initial Setup ###

 * Clone this package (pr2_delivery) into your rosbuild workspace.
 * rosmake pr2_delivery # to build action/message files.
 * Map the environment
 * Copy launch/deliver_server_wg.launch to launch/deliver_server_location_name.launch and fill in the map info for the map you just made.
 * Record poses of the robot base to use for home, get-object, and give-object using test/watch_robot_base_pose.sh and either navigation using rviz or joystick teleop.
 * Copy test/test.py into scripts/do-one-delivery.py or similar and copy the base-pose info into it (x, y, yaw).

### Starting the Action Server ###

 * roslaunch /etc/ros/robot.launch # to bring up the robot
 * roslaunch pr2_delivery deliver_server_location_name.launch
 * *Robot should not immediately do anything*

### Making a Delivery ###

 * rosrun pr2_delivery scripts/do-one-delivery.py
 * *Robot will tuck arms and navigate to get-object location.*
 * *After delivery is done it will tuck arms and drive to home location.*

## Problems that may happen ##

### Hallucinated obstacles ###

The main problem I've had is the tilting laser hitting the
object-being-carried, which causes hallucinations of obstacles about
30 cm in front of the robot.  You should always keep a GridCells
display in RViz showing the local costmap obstacles so you can see if
this is happening.  It's also useful to show a pointcloud of
/tilt_scan_shadow_filtered with Decay Time set to 3 seconds or so.
That is what has the tilt laser data right after the filter.

To fix a problem like this, either modify the tucked_with_object_pose
in DeliverServer.py OR modify the parameters of downscan_filter in
launch/pr2_delivery_nav/tilt_laser_filters.yaml to make it ignore more
of the lower section of the tilt scan data.

The other important thing is that the current numbers have been tuned
for the storage clipboard.  If you use a different object that is
larger, it may trigger this again and need re-tuning.

### Navigating after arm being out ###

I haven't seen this particularly, but when the arm is out to give or
get, the nav stack will think there are obstacles close to the robot
in front.  When the arms tuck, there is some chance that some obstacle
cells in the costmap won't be cleared immediately and the robot will
try to navigate around them.

This can likely be solved by arranging the pick-up and drop-off poses
such that the robot generally turns around to go to the next pose,
instead of needing to drive straight ahead after getting or giving the
object.

### Gripper Cycling ###

There have been a few times where the gripper sensor nodes don't
launch properly.  When that happens, the robot always thinks its hand
is being shaken.  So when it is waiting for an object to be given, it
constantly closes the hand and re-opens it.  If you give it something
in that state, it is bad because when it is time to *give* the object,
it will think someone is pulling on it immediately and thus open the
gripper and drop the object on the floor.

The solution (that I know of) is to re-launch the deliver action
server.  It hasn't happened twice in a row, so this is all I've done
to solve it so far.

### Joystick Teleop Broken ###

I haven't looked into this, but joystick teleop seems to not work when
this is running.  Maybe a mux issue?


CHANGELOG.rst
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Changelog for package pr2_delivery
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

1.0.3 (2014-09-06)
------------------
* Removed makefile
* Re-added proper dependencies
* Contributors: TheDash


test/test.py
#!/usr/bin/env python

import roslib; roslib.load_manifest('pr2_delivery')
import rospy
import actionlib
import geometry_msgs.msg
import tf.transformations
from pr2_delivery.msg import DeliverAction, DeliverGoal

def pose_from_xytheta(x, y, theta):
    pose = geometry_msgs.msg.PoseStamped()
    pose.header.stamp = rospy.Time.now()
    pose.header.frame_id = '/map'
    pose.pose.position.x = x
    pose.pose.position.y = y
    pose.pose.position.z = 0
    quat = tf.transformations.quaternion_from_euler(0, 0, theta)
    pose.pose.orientation = geometry_msgs.msg.Quaternion(*quat)
    return pose

if __name__ == '__main__':
    rospy.loginfo("starting...")
    rospy.init_node('deliver_test_client')
    client = actionlib.SimpleActionClient('deliver', DeliverAction)
    rospy.loginfo("waiting for action server...")
    client.wait_for_server()

    goal = DeliverGoal()
    # Fill in goal here.
    goal.get_object_pose = pose_from_xytheta( 29.494, 31.433, -0.101 )
    goal.give_object_pose = pose_from_xytheta( 27.503, 34.579, 2.464 )
    goal.return_home_pose = pose_from_xytheta( 33.089, 34.547, 3.14 )

    rospy.loginfo("sending goal.")
    client.send_goal(goal)
    client.wait_for_result(rospy.Duration.from_sec(50.0))
    rospy.loginfo("got action finished.")
    rospy.sleep(50)


test/watch_robot_base_pose.sh
#!/bin/bash

# This script is just a convenience for measuring current robot base position and orientation.

rosrun tf tf_echo /map /base_footprint


src/pr2_delivery/Robot.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2013, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import roslib
roslib.load_manifest('pr2_delivery')

import rospy

from trajectory_msgs.msg import *
from actionlib_msgs.msg import *
from pr2_controllers_msgs.msg import *
from pr2_gripper_sensor_msgs.msg import *
import actionlib

# // Our Action interface type, provided as a typedef for convenience
# typedef actionlib::SimpleActionClient<pr2_gripper_sensor_msgs::PR2GripperFindContactAction> ContactClient;
# // Our Action interface type, provided as a typedef for convenience                   
# typedef actionlib::SimpleActionClient<pr2_controllers_msgs::Pr2GripperCommandAction> GripperClient;
# // Our Action interface type, provided as a typedef for convenience
# typedef actionlib::SimpleActionClient<pr2_gripper_sensor_msgs::PR2GripperForceServoAction> ForceClient;


class Robot:
    def __init__(self):
        self.joint_names = ["shoulder_pan", 
                            "shoulder_lift",
                            "upper_arm_roll",
                            "elbow_flex", 
                            "forearm_roll",
                            "wrist_flex", 
                            "wrist_roll" ]

        # Get controller name and start joint trajectory action clients
        self.left_joint_client = actionlib.SimpleActionClient('l_arm_controller/joint_trajectory_action', JointTrajectoryAction)
        self.right_joint_client = actionlib.SimpleActionClient('r_arm_controller/joint_trajectory_action', JointTrajectoryAction)
        self.right_gripper_client = actionlib.SimpleActionClient('r_gripper_sensor_controller/gripper_action', Pr2GripperCommandAction)
        self.right_contact_client = actionlib.SimpleActionClient('r_gripper_sensor_controller/find_contact', PR2GripperFindContactAction)
        self.right_force_client = actionlib.SimpleActionClient('r_gripper_sensor_controller/force_servo', PR2GripperForceServoAction)

        # Wait for joint clients to connect with timeout
        for client in [self.left_joint_client,
                       self.right_joint_client,
                       self.right_gripper_client,
                       self.right_contact_client,
                       self.right_force_client]:
            if not client.wait_for_server(rospy.Duration(30)):
                rospy.logerr("Robot.py: '%s' action server did not come up within timelimit" % client.action_client.ns)

    def move_arm_to(self, side, joint_values, duration):
        """side is 'l' or 'r'.
           joint_values is an array of 7 joint values."""
        goal = JointTrajectoryGoal()
        goal.trajectory.joint_names = [side+"_"+name+"_joint" for name in self.joint_names]
        goal.trajectory.points = []
        goal.trajectory.points.append(JointTrajectoryPoint( positions = joint_values,
                                                            velocities = [],
                                                            accelerations = [],
                                                            time_from_start = rospy.Duration(duration)))
        goal.trajectory.header.stamp = rospy.get_rostime() + rospy.Duration(0.01)
        client = {'l': self.left_joint_client, 'r': self.right_joint_client}[side]
        return client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def open_right_gripper(self):
        goal = Pr2GripperCommandGoal()
        goal.command.position = 0.08
        goal.command.max_effort = -1
        return self.right_gripper_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def grab_with_right_gripper(self):
        """Close gripper until we sense contact on both fingers."""
        goal = PR2GripperFindContactGoal()
        goal.command.contact_conditions = goal.command.BOTH # close until both fingers contact
        goal.command.zero_fingertip_sensors = True # zero fingertip sensor values before moving
        return self.right_contact_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def hold_with_right_gripper(self, hold_force = 10.0):
        """Hold somethign with a constant force (in Newtons) in the gripper"""
        goal = PR2GripperForceServoGoal()
        goal.command.fingertip_force = hold_force
        return self.right_force_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def right_grip(self, position, max_effort = -1):
        goal = Pr2GripperCommandGoal()
        goal.command.position = position
        goal.command.max_effort = max_effort
        return self.right_gripper_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))


src/pr2_delivery/ArmMover.py
#!/usr/bin/env python
# Software License Agreement (BSD License)
#
# Copyright (c) 2013, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of the Willow Garage nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

import roslib
roslib.load_manifest('pr2_delivery')

import rospy

from trajectory_msgs.msg import *
from actionlib_msgs.msg import *
from pr2_controllers_msgs.msg import *
from sensor_msgs.msg import JointState
import actionlib

class Joint:
    def __init__(self, name, position, velocity, effort):
        self.name = name
        self.position = position
        self.velocity = velocity
        self.effort = effort

class ArmMover:
    def __init__(self):
        self.joint_names = ["shoulder_pan", 
                            "shoulder_lift",
                            "upper_arm_roll",
                            "elbow_flex", 
                            "forearm_roll",
                            "wrist_flex", 
                            "wrist_roll" ]

        # Get controller name and start joint trajectory action clients
        self.r_action_name = 'r_arm_controller/joint_trajectory_action'
        self.l_action_name = 'l_arm_controller/joint_trajectory_action'
        self.left_joint_client = actionlib.SimpleActionClient(self.l_action_name, JointTrajectoryAction)
        self.right_joint_client = actionlib.SimpleActionClient(self.r_action_name, JointTrajectoryAction)

        # Wait for joint clients to connect with timeout
        rospy.loginfo("Waiting for " + self.l_action_name)
        if not self.left_joint_client.wait_for_server(rospy.Duration(30)):
            rospy.logerr("ArmMover.py: left_joint_client action server did not come up within timelimit")
        rospy.loginfo("Waiting for " + self.r_action_name)
        if not self.right_joint_client.wait_for_server(rospy.Duration(30)):
            rospy.logerr("ArmMover.py: right_joint_client action server did not come up within timelimit")

        self.r_gripper_action_name = 'r_gripper_controller/gripper_action'
        self.right_gripper_client = actionlib.SimpleActionClient(self.r_gripper_action_name, Pr2GripperCommandAction)

        if not self.right_gripper_client.wait_for_server(rospy.Duration(30)):
            rospy.logerr("ArmMover.py: right_gripper_client action server did not come up within timelimit")

        self.joints = {}
        rospy.Subscriber('/joint_states', JointState, self.joint_states_callback)

    def joint_states_callback(self, msg):
        # string[] name
        # float64[] position
        # float64[] velocity
        # float64[] effort
        i = 0
        for name in msg.name:
            state = Joint(name, msg.position[i], msg.velocity[i], msg.effort[i])
            self.joints[name] = state
            i += 1

    def go(self, side, joint_values, duration):
        """side is 'l' or 'r'.
           joint_values is an array of 7 joint values."""
        goal = JointTrajectoryGoal()
        goal.trajectory.joint_names = [side+"_"+name+"_joint" for name in self.joint_names]
        goal.trajectory.points = []
        goal.trajectory.points.append(JointTrajectoryPoint( positions = joint_values,
                                                            velocities = [],
                                                            accelerations = [],
                                                            time_from_start = rospy.Duration(duration)))
        goal.trajectory.header.stamp = rospy.get_rostime() + rospy.Duration(0.01)
        client = {'l': self.left_joint_client, 'r': self.right_joint_client}[side]
        return client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def open_right(self):
        return self.right_grip(0.08)

    def close_right(self):
        return self.right_grip(0, 200.0)

    def right_grip(self, position, max_effort = -1):
        goal = Pr2GripperCommandGoal()
        goal.command.position = position
        goal.command.max_effort = max_effort
        return self.right_gripper_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def print_arm_pose(self, which_arm):
        joint_pose = []
        for joint in self.joint_names:
            full_name = which_arm + "_" + joint + "_joint"
            joint_pose.append( self.joints[full_name].position )
        print(which_arm, "arm pose", joint_pose)


src/pr2_delivery/__init__.py
# Software License Agreement (BSD License)
#
# Copyright (c) 2013, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Willow Garage, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from .DeliverServer import DeliverServer
__all__ = ['DeliverServer']


src/pr2_delivery/DeliverServer.py
# Software License Agreement (BSD License)
#
# Copyright (c) 2013, Willow Garage, Inc.
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
#
#  * Redistributions of source code must retain the above copyright
#    notice, this list of conditions and the following disclaimer.
#  * Redistributions in binary form must reproduce the above
#    copyright notice, this list of conditions and the following
#    disclaimer in the documentation and/or other materials provided
#    with the distribution.
#  * Neither the name of Willow Garage, Inc. nor the names of its
#    contributors may be used to endorse or promote products derived
#    from this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
# FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
# COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
# INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

PACKAGE='pr2_delivery'

import roslib; roslib.load_manifest(PACKAGE)
import subprocess
import rospy
import actionlib
import geometry_msgs.msg
import pr2_common_action_msgs.msg
from pr2_delivery.msg import DeliverAction, DeliverGoal
from ArmMover import ArmMover
from pr2_gripper_sensor_msgs.msg import PR2GripperEventDetectorAction, PR2GripperEventDetectorGoal
from move_base_msgs.msg import MoveBaseAction, MoveBaseGoal

class DeliverServer:

    tucked_with_object_pose = [-0.0175476818422744, 1.1720448201611564, -1.3268105758514066, -1.288722079574422, -31.28968470078213, -2.0089650532319836, -5.841424529413016]
    tuck_approach_pose = [0.039, 1.1072, 0.0, -2.067, -1.231, -1.998, 0.369]
    accept_object_pose = [ -0.07666010001780543,   0.1622352230632809,   -0.31320771836735584,   -1.374860652847621,   -3.1324415863359545,   -1.078194355846691,   1.857217828689617]
    tucked_with_object_approach_pose = [-0.01829384139848722, 0.6712428753827848, -1.3264898661986668, -0.6078654239376914, 0.601472182148825, -1.3278329090728338, -5.83346239703479]

    def __init__(self):
        self.tuck_arm_client = actionlib.SimpleActionClient("tuck_arms", pr2_common_action_msgs.msg.TuckArmsAction)
        self.gripper_wiggle_detector_client = actionlib.SimpleActionClient('r_gripper_sensor_controller/event_detector', PR2GripperEventDetectorAction)
        self.move_base_client = actionlib.SimpleActionClient('move_base', MoveBaseAction)

        # Load phrases
        self.lang = rospy.get_param('~lang', 'en')
        
        self.request_item_phrase = rospy.get_param('~request_item_phrase', 'Please give me the delivery.')
        self.item_received_phrase = rospy.get_param('~item_received_phrase', 'Thank you. I will deliver this.')
        self.give_item_phrase = rospy.get_param('~give_item_phrase', 'I have a delivery for you.')
        self.item_delivered_phrase = rospy.get_param('~item_delivered_phrase', 'Thank you. Have a nice day.')

        self.arm_mover = ArmMover()

        rospy.loginfo("Waiting for tuck_arms action server")
        self.tuck_arm_client.wait_for_server(rospy.Duration(10.0))
        rospy.loginfo("Waiting for wiggle_detector action server")
        self.gripper_wiggle_detector_client.wait_for_server(rospy.Duration(10.0))
        rospy.loginfo("Waiting for move_base action server")
        self.move_base_client.wait_for_server(rospy.Duration(10.0))

        self.server = actionlib.SimpleActionServer('deliver', DeliverAction, self.execute, False)
        self.server.start()
        
        rospy.loginfo("Ready")

    def execute(self, goal):
        """This is the main sequence of the delivery action."""
        self.tuck_arms()
        self.navigate_to(goal.get_object_pose)
        self.say(self.request_item_phrase)
        self.get_object()
        self.say(self.item_received_phrase)
        self.navigate_to(goal.give_object_pose)
        self.say(self.give_item_phrase)
        self.give_object()
        self.say(self.item_delivered_phrase)
        self.tuck_arms()
        self.navigate_to(goal.return_home_pose)

        self.server.set_succeeded()

    def say(self, thing_to_say):
        #rospy.loginfo("saying '%s'" % thing_to_say)
        # say something
        speed = 130
        subprocess.call([ 'espeak', 
                        "-v", self.lang,
                        "-s", "%d"%speed,
                        thing_to_say ])

    def tuck_arms(self):
        rospy.loginfo("tucking arms")
        goal = pr2_common_action_msgs.msg.TuckArmsGoal()
        goal.tuck_left = True
        goal.tuck_right = True
        self.tuck_arm_client.send_goal_and_wait(goal, rospy.Duration(30.0), rospy.Duration(5.0))

    def navigate_to(self, nav_goal_pose):
        rospy.loginfo("navigating to %f %f" % (nav_goal_pose.pose.position.x, 
                      nav_goal_pose.pose.position.y))
        goal = MoveBaseGoal()
        goal.target_pose = nav_goal_pose
        self.move_base_client.send_goal_and_wait(goal, rospy.Duration(60*5), rospy.Duration(5))

    def wait_for_gripper_wiggle(self, accel):
        """Waits for one year.  accel is in m/s^2, normal values are 6 and 10"""
        # contents of function ported from pr2_props
        goal = PR2GripperEventDetectorGoal()
        goal.command.trigger_conditions = 4 # use just acceleration as our contact signal
        goal.command.acceleration_trigger_magnitude = accel
        goal.command.slip_trigger_magnitude = 0.008 # slip gain
        self.gripper_wiggle_detector_client.send_goal_and_wait(goal, rospy.Duration(3600*24*365), rospy.Duration(5.0))

    def get_object(self):
        rospy.loginfo("getting object")
        # - move right arm to accept-object pose
        self.arm_mover.go('r', self.tuck_approach_pose, 1)
        self.arm_mover.go('r', self.accept_object_pose, 2)

        object_gripped = False
        while not object_gripped:
            # - open gripper all the way
            self.arm_mover.open_right()
            rospy.sleep(2)
            # - wait for externally-applied hand motion detected (ala "fist-pump" demo)
            self.wait_for_gripper_wiggle(10) # m/s^2
            # - close gripper all the way
            self.arm_mover.close_right()
            # - if gripper closes all the way, no object is gripped
            gripper_pos = self.arm_mover.joints['r_gripper_joint'].position
            if gripper_pos > 0.02:
                rospy.loginfo("gripper has object: %f", gripper_pos)
                object_gripped = True
            else:
                rospy.loginfo("gripper does not have object: %f", gripper_pos)

        self.arm_mover.go('r', self.tucked_with_object_approach_pose, 2)
        self.arm_mover.go('r', self.tucked_with_object_pose, 1)

    def give_object(self):
        rospy.loginfo("giving object")
        # move out to tucked-with-object approach pose for right arm
        self.arm_mover.go('r', self.tucked_with_object_approach_pose, 2)
        # - move right arm to give-object pose
        self.arm_mover.go('r', self.accept_object_pose, 2) # give-object pose is the same as accept-object pose.
        # - let arm motion settle
        rospy.sleep(2) 
        # - wait for externally-applied hand motion detected (ala "fist-pump" demo)
        self.wait_for_gripper_wiggle(5) # m/s^2
        # - open gripper
        self.arm_mover.open_right()
        rospy.sleep(1) 

# Question: what if sequence is pre-empted while robot is holding object?
# - maybe set the object on the floor?
# - v1 is certainly just "do nothing" so robot is still holding the object.


scripts/call_deliver.py
#!/usr/bin/env python

import roslib; roslib.load_manifest('pr2_delivery')
import rospy
import actionlib
import geometry_msgs.msg
import tf.transformations
from pr2_delivery.msg import DeliverAction, DeliverGoal

import yaml
import os

def pose_from_yaml(filename):
    stream = open( filename )
    pose_dict = yaml.load( stream )
    
    pose = geometry_msgs.msg.PoseStamped()
    pose.header.stamp = rospy.Time.now()
    pose.header.frame_id = '/map'
    
    pose.pose.position.x = pose_dict['pose']['position']['x']
    pose.pose.position.y = pose_dict['pose']['position']['y']
    pose.pose.position.z = pose_dict['pose']['position']['z']
    
    pose.pose.orientation.x = pose_dict['pose']['orientation']['x']
    pose.pose.orientation.y = pose_dict['pose']['orientation']['y']
    pose.pose.orientation.z = pose_dict['pose']['orientation']['z']
    pose.pose.orientation.w = pose_dict['pose']['orientation']['w']
    
    return pose

if __name__ == '__main__':
    rospy.loginfo("starting...")
    rospy.init_node('deliver_test_client')
    client = actionlib.SimpleActionClient('deliver', DeliverAction)
    rospy.loginfo("waiting for action server...")
    client.wait_for_server()

    data_dir = rospy.get_param("~data_dir")
    rospy.loginfo("Using %s as data directory"%(data_dir))

    goal = DeliverGoal()
    # Fill in goal here.
    goal.get_object_pose = pose_from_yaml(os.path.join(data_dir,'get_object_pose.yaml'))
    goal.give_object_pose = pose_from_yaml(os.path.join(data_dir,'give_object_pose.yaml'))
    goal.return_home_pose = pose_from_yaml(os.path.join(data_dir,'return_home_pose.yaml'))

    rospy.loginfo("sending goal.")
    client.send_goal(goal)
    client.wait_for_result(rospy.Duration.from_sec(50.0))
    rospy.loginfo("got action finished.")
    rospy.sleep(50)
